src/main/java/org/jabref/gui/preferences/PreferencesSearchHandler.java

```diff
@@ -0,0 +1,140 @@
+
+class PreferencesSearchHandler {
+
+    private static PseudoClass labelHighlight = PseudoClass.getPseudoClass("search-highlight");
+    private final List<PrefsTab> preferenceTabs;
+    private final StringProperty searchText;
+    private final ListProperty<PrefsTab> filteredPreferenceTabs;
+    private final ArrayListMultimap<PrefsTab, LabeledWrapper> preferenceTabsLabelNames;
+    private final ArrayList<LabeledWrapper> highlightedLabels = new ArrayList<>();
+
+    private class LabeledWrapper {
+
+        private final Labeled labeled;
+
+        LabeledWrapper(Labeled _labeled) {
+            labeled = _labeled;
+        }
+
+        public boolean contains(String query) {
+            return labeled.getText().toLowerCase(Locale.ROOT).contains(query.toLowerCase(Locale.ROOT));
+        }
+
+        public String getText() {
+            return labeled.getText();
+        }
+
+        public void setHighlighted(boolean highlight) {
+            labeled.pseudoClassStateChanged(labelHighlight, highlight);
+        }
+    }
+
+    PreferencesSearchHandler(List<PrefsTab> preferenceTabs, StringProperty searchText) {
+        this.preferenceTabs = preferenceTabs;
+        this.searchText = searchText;
+        this.preferenceTabsLabelNames = getPrefsTabLabelMap();
+        this.filteredPreferenceTabs = new SimpleListProperty<>(FXCollections.observableArrayList(preferenceTabs));
+        initializeSearchTextListener();
+    }
+
+    /**
+     * Reacts upon changes in the search text.
+     */
+    private void initializeSearchTextListener() {
+        searchText.addListener((observable, previousSearchText, newSearchText) -> {
+            clearHighlights();
+            if (newSearchText.isEmpty()) {
+                clearSearch();
+            } else {
+                updateSearch(newSearchText);
+            }
+        });
+    }
+
+    private void updateSearch(String newSearchText) {
+        filterByTabName(newSearchText);
+        filterByLabelName(newSearchText);
+    }
+
+    private void filterByTabName(String newSearchText) {
+        List<PrefsTab> filteredTabs = preferenceTabs.stream()
+                                                    .filter(tab -> tab.getTabName().toLowerCase(Locale.ROOT).contains(newSearchText))
+                                                    .collect(Collectors.toCollection(ArrayList::new));
+        filteredPreferenceTabs.setAll(filteredTabs);
+    }
+
+    private void filterByLabelName(String newSearchText) {
+        for (PrefsTab tab : preferenceTabsLabelNames.keySet()) {
+            // If the current tab contains a matching label
+            boolean tabContainsLabel = false;
+            for (LabeledWrapper labeledWrapper : preferenceTabsLabelNames.get(tab)) {
+                if (labeledWrapper.contains(newSearchText)) {
+                    tabContainsLabel = true;
+                    highlightLabel(labeledWrapper);
+                }
+            }
+
+            if (tabContainsLabel) {
```

Comment:
As you suggested yourself, I think it is a good idea to move the logic of `filterByTabName` here (`if (tabContainsLabel || tabNameIsMatchedBySearchQuery`) instead of having two separate methods.